src/librustc_typeck/check/mod.rs

```diff
@@ -1706,11 +1690,98 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     }
 
     fn select_all_obligations_and_apply_defaults(&self) {
-        debug!("select_all_obligations_and_apply_defaults");
+        use middle::ty::UnconstrainedNumeric::{UnconstrainedInt, UnconstrainedFloat, Neither};
 
-        self.select_obligations_where_possible();
-        self.default_type_parameters();
-        self.select_obligations_where_possible();
+        debug!("select_all_obligations_and_apply_defaults: defaults={:?}", self.infcx().defaults);
+
+        for _ in (0..self.tcx().sess.recursion_limit.get()) {
+            self.select_obligations_where_possible();
+
+            let unsolved_variables = self.inh.infcx.unsolved_variables();
+            let mut unbound_tyvars = HashSet::new();
+
+            // Gather all unconstrainted integer and float variables
+            for ty in &unsolved_variables {
+                let resolved = self.infcx().resolve_type_vars_if_possible(ty);
+                if self.infcx().type_var_diverges(resolved) {
+                    demand::eqtype(self, codemap::DUMMY_SP, *ty, self.tcx().mk_nil());
+                } else {
+                    match self.infcx().type_is_unconstrained_numeric(resolved) {
+                        UnconstrainedInt => {
+                            unbound_tyvars.insert(resolved);
+                        },
+                        UnconstrainedFloat => {
+                            unbound_tyvars.insert(resolved);
+                        }
+                        Neither => {}
+                    }
+                }
+            }
+
+            // Collect the set of variables that need fallback applied
+            for ty in &unsolved_variables {
+                if self.inh.infcx.defaults.borrow().contains_key(ty) {
+                    let resolved = self.infcx().resolve_type_vars_if_possible(ty);
+
+                    debug!("select_all_obligations_and_apply_defaults: ty: {:?} with default: {:?}",
+                            ty, self.inh.infcx.defaults.borrow().get(ty));
+
+                    match resolved.sty {
+                        ty::TyInfer(ty::TyVar(_)) => {
+                            unbound_tyvars.insert(ty);
+                        }
+
+                        ty::TyInfer(ty::IntVar(_)) | ty::TyInfer(ty::FloatVar(_)) => {
+                            unbound_tyvars.insert(ty);
+                            if unbound_tyvars.contains(resolved) {
+                                unbound_tyvars.remove(resolved);
+                            }
+                        }
+
+                        _ => {}
+                    }
+                }
+            }
+
+            if unbound_tyvars.is_empty() {
+                break;
+            }
+
+            // Go through the unbound variables and unify them with the proper fallbacks
+            for ty in &unbound_tyvars {
+                // let resolved = self.infcx().resolve_type_vars_if_possible(ty);
+                if self.infcx().type_var_diverges(ty) {
```

Comment:
It is my memory that I tried to remove this and then couldn't compile libcore, it would be nice to get rid of of ...